velox/dwio/dwrf/test/CacheInputTest.cpp

```diff
@@ -852,3 +868,66 @@ TEST_F(CacheTest, loadQuotumTooLarge) {
           readOptions),
       "Load quantum exceeded SSD cache entry size limit");
 }
+
+TEST_F(CacheTest, ssdReadVerification) {
+  constexpr int64_t kMemoryBytes = 32 << 20;
+  constexpr int64_t kSsdBytes = 256 << 20;
+  // 32 RAM, 256MB SSD, with checksumWrite/checksumReadVerification enabled.
+  initializeCache(kMemoryBytes, kSsdBytes, true);
+
+  uint64_t fileId;
+  uint64_t groupId;
+  auto file = inputByPath("test_file", fileId, groupId);
+  auto tracker = std::make_shared<ScanTracker>(
+      "testTracker", nullptr, io::ReaderOptions::kDefaultLoadQuantum);
+  auto input = std::make_unique<CachedBufferedInput>(
+      file,
+      MetricsLog::voidLog(),
+      fileId,
+      cache_.get(),
+      tracker,
+      groupId,
+      ioStats_,
+      executor_.get(),
+      io::ReaderOptions(pool_.get()));
+
+  const auto readData = [&](uint32_t numBytesRead) {
+    const uint64_t kNumBytesPerRead = 8 << 20;
+    for (uint64_t offset = 0; offset < numBytesRead;
+         offset += kNumBytesPerRead) {
+      auto stream = input->read(offset, kNumBytesPerRead, LogType::TEST);
+      const void* buffer;
+      int32_t size;
+      int32_t bytes = 0;
+      while (bytes < kNumBytesPerRead) {
+        EXPECT_TRUE(stream->Next(&buffer, &size));
+        bytes += size;
+      }
+    }
+  };
+
+  // Read kSsdBytes of data.
+  readData(kSsdBytes);
+  auto stats = cache_->refreshStats();
+  // This is a cold read, so expect no cache hit.
+  ASSERT_EQ(stats.numHit, 0);
+  ASSERT_EQ(stats.ssdStats->entriesRead, 0);
+  ASSERT_EQ(stats.ssdStats->readSsdCorruptions, 0);
+
+  // Read kSsdBytes of data.
+  readData(kSsdBytes);
+  stats = cache_->refreshStats();
+  // Expect cache hits.
+  ASSERT_GT(stats.numHit, 0);
+  ASSERT_GT(stats.ssdStats->entriesRead, 0);
+  ASSERT_EQ(stats.ssdStats->readSsdCorruptions, 0);
+
+  // Corrupt SSD cache file.
+  corruptSsdFile(fmt::format("{}/cache0", tempDirectory_->getPath()));
+
+  // Read from the corrupted cache.
+  readData(kSsdBytes);
```

Comment:
Can you add to check storage read stats in io stats?  Check where ssd and storage read counts are updated in CacheInputStream::loadSync? Thanks!